File: practice-cpp/fundamentals/src/bucles.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {

    // ===== FOR CLÁSICO =====

    cout << "\n===== FOR CLÁSICO =====\n";

    // estructura: for (inicialización; condición; incremento)
    for (int i = 1; i <= 5; i++) {
        cout << "Iteración " << i << "\n";
    }

    cout << "\nTabla de multiplicar del 3:\n";
    for (int i = 1; i <= 10; i++) {
        cout << "3 * " << i << " = " << (3 * i) << "\n";
    }

    // ===== FOR CON DECREMENTO =====

    cout << "\n===== FOR CON DECREMENTO =====\n";

    for (int i = 5; i >= 1; i--) {
        cout << "Cuenta atrás: " << i << "\n";
    }

    // ===== WHILE =====

    cout << "\n===== WHILE =====\n";

    // estructura: while (condición) { ... }
    // se evalúa la condición primero, luego ejecuta el cuerpo
    int contador = 1;
    cout << "Números del 1 al 5 con while:\n";
    while (contador <= 5) {
        cout << contador << "\n";
        contador++;
    }

    // ===== DO-WHILE =====

    cout << "\n===== DO-WHILE =====\n";

    // estructura: do { ... } while (condición)
    // ejecuta el cuerpo primero, luego evalúa la condición
    // garantiza al menos una ejecución
    int contadorDo = 1;
    cout << "Números del 1 al 5 con do-while:\n";
    do {
        cout << contadorDo << "\n";
        contadorDo++;
    } while (contadorDo <= 5);

    // ===== RANGE-BASED FOR (C++11) =====

    cout << "\n===== RANGE-BASED FOR (C++11) =====\n";

    // estructura: for (tipo variable : colección) { ... }
    // útil para recorrer vectores y arrays sin índice
    vector<int> numeros = {10, 20, 30, 40, 50};
    cout << "Recorrido de vector con range-based for:\n";
    for (int numero : numeros) {
        cout << "Elemento: " << numero << "\n";
    }

    // ===== BREAK: SALIR DEL BUCLE =====

    cout << "\n===== BREAK: SALIR DEL BUCLE =====\n";

    cout << "Conteo hasta encontrar 5:\n";
    for (int i = 1; i <= 10; i++) {
        cout << i << "\n";
        if (i == 5) {
            cout << "¡Encontrado el 5! Saliendo del bucle.\n";
            break; // sale del bucle inmediatamente
        }
    }

    // ===== CONTINUE: SALTAR A LA SIGUIENTE ITERACIÓN =====

    cout << "\n===== CONTINUE: SALTAR ITERACIÓN =====\n";

    cout << "Números impares del 1 al 10:\n";
    for (int i = 1; i <= 10; i++) {
        if (i % 2 == 0) {
            continue; // salta el resto del código en esta iteración
        }
        cout << i << "\n";
    }

    // ===== BUCLES ANIDADOS =====

    cout << "\n===== BUCLES ANIDADOS =====\n";

    cout << "Tabla de multiplicar (2x2 a 5x5):\n";
    for (int i = 2; i <= 5; i++) {
        cout << "\nTabla del " << i << ":\n";
        for (int j = 1; j <= 5; j++) {
            cout << "  " << i << " * " << j << " = " << (i * j) << "\n";
        }
    }

    // ===== SUMA ACUMULATIVA CON BUCLE =====

    cout << "\n===== SUMA ACUMULATIVA CON BUCLE =====\n";

    int suma = 0;
    int n = 5;
    for (int i = 1; i <= n; i++) {
        suma += i; // suma acumulativa: suma = suma + i
        cout << "Iteración " << i << ", suma acumulada: " << suma << "\n";
    }
    cout << "Suma total de 1 a " << n << ": " << suma << "\n";

    // ===== BÚSQUEDA EN VECTOR CON BREAK =====

    cout << "\n===== BÚSQUEDA EN VECTOR =====\n";

    vector<int> valores = {100, 200, 300, 400, 500};
    int buscar = 300;
    bool encontrado = false;

    for (int valor : valores) {
        if (valor == buscar) {
            cout << "¡Encontrado " << buscar << " en el vector!\n";
            encontrado = true;
            break;
        }
    }

    if (!encontrado) {
        cout << "El valor " << buscar << " no está en el vector\n";
    }

    // ===== PRODUCTO ACUMULATIVO CON BUCLE =====

    cout << "\n===== PRODUCTO ACUMULATIVO (FACTORIAL) =====\n";

    int producto = 1;
    for (int i = 1; i <= 4; i++) {
        producto *= i; // factorial: 1 * 2 * 3 * 4
    }
    cout << "Factorial de 4: " << producto << "\n";

    return 0;
}

File: practice-rust/fundamentals/src/bucles.rs
// ===== BUCLES EN RUST =====

fn main() {
    println!("\n===== BUCLE LOOP (INFINITO) =====");

    let mut contador = 0;
    loop {
        println!("Contador: {}", contador);
        contador += 1;

        if contador >= 3 {
            break;  // Salir del bucle
        }
    }

    println!("\n===== LOOP CON BREAK Y VALOR =====");

    let mut contador = 0;
    let resultado = loop {
        contador += 1;
        if contador == 5 {
            break contador * 2;  // Retornar valor
        }
    };
    println!("Resultado del loop: {}", resultado);

    println!("\n===== WHILE =====");

    let mut numero = 1;
    println!("Números impares hasta 10:");
    while numero <= 10 {
        if numero % 2 != 0 {
            print!("{} ", numero);
        }
        numero += 1;
    }
    println!();

    println!("\n===== FOR CON RANGO =====");

    println!("Números del 1 al 5:");
    for i in 1..=5 {
        print!("{} ", i);
    }
    println!();

    println!("Números del 1 al 4 (sin incluir 5):");
    for i in 1..5 {
        print!("{} ", i);
    }
    println!();

    println!("\n===== FOR CON ARRAY =====");

    let array = [10, 20, 30, 40, 50];
    println!("Elementos del array: ");
    for elemento in array {
        println!("  {}", elemento);
    }

    println!("\n===== FOR CON ÍNDICE =====");

    let valores = vec![100, 200, 300, 400];
    for (indice, valor) in valores.iter().enumerate() {
        println!("valores[{}] = {}", indice, valor);
    }

    println!("\n===== FOR CON MUTABLE REFERENCE =====");

    let mut numeros = vec![1, 2, 3, 4, 5];
    for numero in &mut numeros {
        *numero *= 2;  // Modificar cada elemento
    }
    println!("Números duplicados: {:?}", numeros);

    println!("\n===== BREAK Y CONTINUE =====");

    println!("Números del 1 al 10 (saltando pares):");
    for i in 1..=10 {
        if i % 2 == 0 {
            continue;  // Saltar a siguiente iteración
        }
        print!("{} ", i);
    }
    println!();

    println!("Números hasta encontrar 5:");
    for i in 1..=10 {
        print!("{} ", i);
        if i == 5 {
            break;  // Salir del bucle
        }
    }
    println!();

    println!("\n===== BUCLES ANIDADOS =====");

    println!("Tabla de multiplicar 3x3:");
    for i in 1..=3 {
        for j in 1..=3 {
            print!("({},{}) ", i, j);
        }
        println!();
    }

    println!("\n===== ETIQUETAS DE BUCLE =====");

    let mut contador = 0;
    'bucle_externo: loop {
        let mut contador_interno = 0;
        loop {
            contador_interno += 1;
            contador += 1;

            if contador_interno == 3 {
                break;  // Salir del bucle interno
            }

            if contador == 5 {
                break 'bucle_externo;  // Salir del bucle externo
            }
        }
    }
    println!("Contador final: {}", contador);

    println!("\n===== SUMA ACUMULATIVA =====");

    let mut suma = 0;
    for i in 1..=5 {
        suma += i;
        println!("Suma hasta {}: {}", i, suma);
    }

    println!("\n===== BÚSQUEDA CON BREAK =====");

    let numeros = vec![10, 20, 30, 40, 50];
    let buscado = 30;
    let mut encontrado = false;

    for numero in &numeros {
        if *numero == buscado {
            println!("Número {} encontrado", buscado);
            encontrado = true;
            break;
        }
    }

    if !encontrado {
        println!("Número {} no encontrado", buscado);
    }

    println!("\n===== ITERADORES (más Rust) =====");

    let numeros = vec![1, 2, 3, 4, 5];

    // using iter()
    println!("Con iter():");
    for num in numeros.iter() {
        print!("{} ", num);
    }
    println!();

    // usando into_iter() - consume el vector
    println!("Con into_iter():");
    let numeros2 = vec![10, 20, 30];
    for num in numeros2 {
        print!("{} ", num);
    }
    println!();

    println!("\n===== RESUMEN DE BUCLES =====");

    println!("loop - bucle infinito (con break)");
    println!("while - mientras la condición sea true");
    println!("for - iterar sobre colecciones o rangos");
    println!("for in range - 1..5 (sin incluir 5), 1..=5 (incluye 5)");
    println!("break - salir del bucle");
    println!("continue - saltar a la siguiente iteración");
    println!("Etiquetas - para controlar bucles anidados");
}
